### DISCUSSION TITLE
DB, cache, and s3 storage from tighten/takeout

### DISCUSSION THREAD

**Discussion by larskoole** [Vessel-App/vessel-cli#16] [Ideas]
I have no idea if this is even remotely possible but it would be nice if we could leverage [tighten/takeout](https://github.com/tighten/takeout) to take care of DB, cache and S3 compatible storage.

Maybe automatically load their docker images from their repo or something 🤷 

Anyways, very cool package. I really like its simplicity and it seems like this could be a product on its own. (looking at the way Jetbrains and Microsoft are already getting IDEs to the cloud, why not the whole dev environment right?)

**Comment by fideloper** [comment]
PHP is being run on the remote server, so the remote server would need to be able to communicate back to your local instance. I don’t think that’s really possible (or rather, i don’t know how - seems like it should be possible)! But i like that idea!

The plan right now is to have this tool spin up another VM for you, for redis, database, etc. preferably it would manage turning those on and off as well!

**Reply by larskoole** [reply]
Oh yea what I meant was running takeout on a Fly Machine. But I have no idea whether that would be efficient or even possible.

But a solution with less overhead would probably be better yea😅

Just hoping to see some Mysql or SingleStore cluster-in-a-box support in the future. I could even send in a PR once there's a base template😄

Can't stop being hyped about this!

**Reply by fideloper** [reply]
Right on! So there's a fly article on running mysql: https://fly.io/docs/app-guides/mysql-on-fly/

Basically you just take the mysql official Docker image and make it work (just like Takeout). 

Those docs are a bit specific to Fly Apps, however this uses [Fly machines](https://fly.io/docs/reference/machines/). But all the concepts apply - you tell the Machine to run the MySQL official Docker container and make a VM from it.

Vessel needs tweaks to manage the lifecycle of of that (create it, turn it off, turn it on, etc) and to be able to create a volume, so data persists.

We'll also need a way to track "state" - if a VM exists already or not. That can likely be some data stored in `~/.vessel/envs/<some-app>/some_file.ext`
